File: internal/websocket/handler/control.go
package handler

import (
	"errors"

	"github.com/demodesk/neko/pkg/types"
	"github.com/demodesk/neko/pkg/types/event"
	"github.com/demodesk/neko/pkg/types/message"
	"github.com/demodesk/neko/pkg/xorg"
)

var (
	ErrIsNotAllowedToHost = errors.New("is not allowed to host")
	ErrIsNotTheHost       = errors.New("is not the host")
	ErrIsAlreadyTheHost   = errors.New("is already the host")
	ErrIsAlreadyHosted    = errors.New("is already hosted")
)

func (h *MessageHandlerCtx) controlRelease(session types.Session) error {
	if !session.Profile().CanHost || session.PrivateModeEnabled() {
		return ErrIsNotAllowedToHost
	}

	if !session.IsHost() {
		return ErrIsNotTheHost
	}

	h.desktop.ResetKeys()
	h.sessions.ClearHost()

	return nil
}

func (h *MessageHandlerCtx) controlRequest(session types.Session) error {
	if !session.Profile().CanHost || session.PrivateModeEnabled() {
		return ErrIsNotAllowedToHost
	}

	if session.IsHost() {
		return ErrIsAlreadyTheHost
	}

	if h.sessions.Settings().LockedControls && !session.Profile().IsAdmin {
		return ErrIsNotAllowedToHost
	}

	if !h.sessions.Settings().ImplicitHosting {
		// tell session if there is a host
		if host, hasHost := h.sessions.GetHost(); hasHost {
			session.Send(
				event.CONTROL_HOST,
				message.ControlHost{
					HasHost: true,
					HostID:  host.ID(),
				})

			return ErrIsAlreadyHosted
		}
	}

	h.sessions.SetHost(session)

	return nil
}

func (h *MessageHandlerCtx) controlMove(session types.Session, payload *message.ControlPos) error {
	if err := h.controlRequest(session); err != nil && !errors.Is(err, ErrIsAlreadyTheHost) {
		return err
	}

	// handle active cursor movement
	h.desktop.Move(payload.X, payload.Y)
	h.webrtc.SetCursorPosition(payload.X, payload.Y)
	return nil
}

func (h *MessageHandlerCtx) controlScroll(session types.Session, payload *message.ControlScroll) error {
	if err := h.controlRequest(session); err != nil && !errors.Is(err, ErrIsAlreadyTheHost) {
		return err
	}

	// TOOD: remove this once the client is fixed
	if payload.DeltaX == 0 && payload.DeltaY == 0 {
		payload.DeltaX = payload.X
		payload.DeltaY = payload.Y
	}

	h.desktop.Scroll(payload.DeltaX, payload.DeltaY, payload.ControlKey)
	return nil
}

func (h *MessageHandlerCtx) controlButtonPress(session types.Session, payload *message.ControlButton) error {
	if payload.ControlPos != nil {
		if err := h.controlMove(session, payload.ControlPos); err != nil {
			return err
		}
	} else if err := h.controlRequest(session); err != nil && !errors.Is(err, ErrIsAlreadyTheHost) {
		return err
	}

	return h.desktop.ButtonPress(payload.Code)
}

func (h *MessageHandlerCtx) controlButtonDown(session types.Session, payload *message.ControlButton) error {
	if payload.ControlPos != nil {
		if err := h.controlMove(session, payload.ControlPos); err != nil {
			return err
		}
	} else if err := h.controlRequest(session); err != nil && !errors.Is(err, ErrIsAlreadyTheHost) {
		return err
	}

	return h.desktop.ButtonDown(payload.Code)
}

func (h *MessageHandlerCtx) controlButtonUp(session types.Session, payload *message.ControlButton) error {
	if payload.ControlPos != nil {
		if err := h.controlMove(session, payload.ControlPos); err != nil {
			return err
		}
	} else if err := h.controlRequest(session); err != nil && !errors.Is(err, ErrIsAlreadyTheHost) {
		return err
	}

	return h.desktop.ButtonUp(payload.Code)
}

func (h *MessageHandlerCtx) controlKeyPress(session types.Session, payload *message.ControlKey) error {
	if payload.ControlPos != nil {
		if err := h.controlMove(session, payload.ControlPos); err != nil {
			return err
		}
	} else if err := h.controlRequest(session); err != nil && !errors.Is(err, ErrIsAlreadyTheHost) {
		return err
	}

	return h.desktop.KeyPress(payload.Keysym)
}

func (h *MessageHandlerCtx) controlKeyDown(session types.Session, payload *message.ControlKey) error {
	if payload.ControlPos != nil {
		if err := h.controlMove(session, payload.ControlPos); err != nil {
			return err
		}
	} else if err := h.controlRequest(session); err != nil && !errors.Is(err, ErrIsAlreadyTheHost) {
		return err
	}

	return h.desktop.KeyDown(payload.Keysym)
}

func (h *MessageHandlerCtx) controlKeyUp(session types.Session, payload *message.ControlKey) error {
	if payload.ControlPos != nil {
		if err := h.controlMove(session, payload.ControlPos); err != nil {
			return err
		}
	} else if err := h.controlRequest(session); err != nil && !errors.Is(err, ErrIsAlreadyTheHost) {
		return err
	}

	return h.desktop.KeyUp(payload.Keysym)
}

func (h *MessageHandlerCtx) controlTouchBegin(session types.Session, payload *message.ControlTouch) error {
	if err := h.controlRequest(session); err != nil && !errors.Is(err, ErrIsAlreadyTheHost) {
		return err
	}
	return h.desktop.TouchBegin(payload.TouchId, payload.X, payload.Y, payload.Pressure)
}

func (h *MessageHandlerCtx) controlTouchUpdate(session types.Session, payload *message.ControlTouch) error {
	if err := h.controlRequest(session); err != nil && !errors.Is(err, ErrIsAlreadyTheHost) {
		return err
	}
	return h.desktop.TouchUpdate(payload.TouchId, payload.X, payload.Y, payload.Pressure)
}

func (h *MessageHandlerCtx) controlTouchEnd(session types.Session, payload *message.ControlTouch) error {
	if err := h.controlRequest(session); err != nil && !errors.Is(err, ErrIsAlreadyTheHost) {
		return err
	}
	return h.desktop.TouchEnd(payload.TouchId, payload.X, payload.Y, payload.Pressure)
}

func (h *MessageHandlerCtx) controlCut(session types.Session) error {
	if err := h.controlRequest(session); err != nil && !errors.Is(err, ErrIsAlreadyTheHost) {
		return err
	}

	return h.desktop.KeyPress(xorg.XK_Control_L, xorg.XK_x)
}

func (h *MessageHandlerCtx) controlCopy(session types.Session) error {
	if err := h.controlRequest(session); err != nil && !errors.Is(err, ErrIsAlreadyTheHost) {
		return err
	}

	return h.desktop.KeyPress(xorg.XK_Control_L, xorg.XK_c)
}

func (h *MessageHandlerCtx) controlPaste(session types.Session, payload *message.ClipboardData) error {
	if err := h.controlRequest(session); err != nil && !errors.Is(err, ErrIsAlreadyTheHost) {
		return err
	}

	// if there have been set clipboard data, set them first
	if payload != nil && payload.Text != "" {
		if err := h.clipboardSet(session, payload); err != nil {
			return err
		}
	}

	return h.desktop.KeyPress(xorg.XK_Control_L, xorg.XK_v)
}

func (h *MessageHandlerCtx) controlSelectAll(session types.Session) error {
	if err := h.controlRequest(session); err != nil && !errors.Is(err, ErrIsAlreadyTheHost) {
		return err
	}

	return h.desktop.KeyPress(xorg.XK_Control_L, xorg.XK_a)
}

File: src/component/internal/control.ts
import * as EVENT from '../types/events'
import type * as message from '../types/messages'

import EventEmitter from 'eventemitter3'
import type { NekoConnection } from './connection'
import type { Control } from '../types/state'

export interface NekoControlEvents {
  ['overlay.click']: (e: MouseEvent) => void
  ['overlay.contextmenu']: (e: MouseEvent) => void
}

export interface ControlPos {
  x: number
  y: number
}

export interface ControlScroll {
  delta_x: number
  delta_y: number
  control_key?: boolean
}

export class NekoControl extends EventEmitter<NekoControlEvents> {
  // eslint-disable-next-line
  constructor(
    private readonly _connection: NekoConnection,
    private readonly _state: Control,
  ) {
    super()
  }

  get useWebrtc() {
    // we want to use webrtc if we're connected and we're the host
    // because webrtc is faster and it doesn't request control
    // in contrast to the websocket
    return this._connection.webrtc.connected && this._state.is_host
  }

  get enabledTouchEvents() {
    return this._state.touch.enabled
  }

  get supportedTouchEvents() {
    return this._state.touch.supported
  }

  public lock() {
    this._state.locked = true // TODO: Vue.Set
  }

  public unlock() {
    this._state.locked = false // TODO: Vue.Set
  }

  public request() {
    this._connection.websocket.send(EVENT.CONTROL_REQUEST)
  }

  public release() {
    this._connection.websocket.send(EVENT.CONTROL_RELEASE)
  }

  public move(pos: ControlPos) {
    if (this.useWebrtc) {
      this._connection.webrtc.send('mousemove', pos)
    } else {
      this._connection.websocket.send(EVENT.CONTROL_MOVE, pos as message.ControlPos)
    }
  }

  // TODO: add pos parameter
  public scroll(scroll: ControlScroll) {
    if (this.useWebrtc) {
      this._connection.webrtc.send('wheel', scroll)
    } else {
      this._connection.websocket.send(EVENT.CONTROL_SCROLL, scroll as message.ControlScroll)
    }
  }

  // buttonpress ensures that only one button is pressed at a time
  public buttonPress(code: number, pos?: ControlPos) {
    this._connection.websocket.send(EVENT.CONTROL_BUTTONPRESS, { code, ...pos } as message.ControlButton)
  }

  public buttonDown(code: number, pos?: ControlPos) {
    if (this.useWebrtc) {
      if (pos) this._connection.webrtc.send('mousemove', pos)
      this._connection.webrtc.send('mousedown', { key: code })
    } else {
      this._connection.websocket.send(EVENT.CONTROL_BUTTONDOWN, { code, ...pos } as message.ControlButton)
    }
  }

  public buttonUp(code: number, pos?: ControlPos) {
    if (this.useWebrtc) {
      if (pos) this._connection.webrtc.send('mousemove', pos)
      this._connection.webrtc.send('mouseup', { key: code })
    } else {
      this._connection.websocket.send(EVENT.CONTROL_BUTTONUP, { code, ...pos } as message.ControlButton)
    }
  }

  // keypress ensures that only one key is pressed at a time
  public keyPress(keysym: number, pos?: ControlPos) {
    this._connection.websocket.send(EVENT.CONTROL_KEYPRESS, { keysym, ...pos } as message.ControlKey)
  }

  public keyDown(keysym: number, pos?: ControlPos) {
    if (this.useWebrtc) {
      if (pos) this._connection.webrtc.send('mousemove', pos)
      this._connection.webrtc.send('keydown', { key: keysym })
    } else {
      this._connection.websocket.send(EVENT.CONTROL_KEYDOWN, { keysym, ...pos } as message.ControlKey)
    }
  }

  public keyUp(keysym: number, pos?: ControlPos) {
    if (this.useWebrtc) {
      if (pos) this._connection.webrtc.send('mousemove', pos)
      this._connection.webrtc.send('keyup', { key: keysym })
    } else {
      this._connection.websocket.send(EVENT.CONTROL_KEYUP, { keysym, ...pos } as message.ControlKey)
    }
  }

  public touchBegin(touch_id: number, pos: ControlPos, pressure: number) {
    if (this.useWebrtc) {
      this._connection.webrtc.send('touchbegin', { touch_id, ...pos, pressure })
    } else {
      this._connection.websocket.send(EVENT.CONTROL_TOUCHBEGIN, { touch_id, ...pos, pressure } as message.ControlTouch)
    }
  }

  public touchUpdate(touch_id: number, pos: ControlPos, pressure: number) {
    if (this.useWebrtc) {
      this._connection.webrtc.send('touchupdate', { touch_id, ...pos, pressure })
    } else {
      this._connection.websocket.send(EVENT.CONTROL_TOUCHUPDATE, { touch_id, ...pos, pressure } as message.ControlTouch)
    }
  }

  public touchEnd(touch_id: number, pos: ControlPos, pressure: number) {
    if (this.useWebrtc) {
      this._connection.webrtc.send('touchend', { touch_id, ...pos, pressure })
    } else {
      this._connection.websocket.send(EVENT.CONTROL_TOUCHEND, { touch_id, ...pos, pressure } as message.ControlTouch)
    }
  }

  public cut() {
    this._connection.websocket.send(EVENT.CONTROL_CUT)
  }

  public copy() {
    this._connection.websocket.send(EVENT.CONTROL_COPY)
  }

  public paste(text?: string) {
    this._connection.websocket.send(EVENT.CONTROL_PASTE, { text } as message.ClipboardData)
  }

  public selectAll() {
    this._connection.websocket.send(EVENT.CONTROL_SELECT_ALL)
  }
}
